Miscellaneous support for cgap-portal, and some unit testing (part of C4-601)

netsettler opened the pull request:
For `jh_utils`:

* I wrote a unit test for `open_4dn_file` that can replace the integration test (which I've marked `integrationx` so it won't run normally among the unit tests). This fixes one of the several failing tests mentioned in [Several unit tests in dcicutils are failing when nothing has changed (C4-601)](https://hms-dbmi.atlassian.net/browse/C4-601).
* I wrote a new unit test for `find_valid_file_or_extra_file`.

For `misc_utils`:

* New function `ignorable` which is basically a synonym for `ignore`, but with the sense that it's OK for the variables given as its arguments to be used elsewhere or not.
* New function `ancestor_classes` that returns a list of the classes from which a given class inherits.
* New function `is_proper_subclass` which is like `issubclass` but returns true only if its two arguments are not the same.
* New function `identity` that returns its argument.
* New functions `count` and `count_if` for counting things in a sequence.
* New function `find_association` for finding 
* New `@decorator` decorator for defining--what else?--decorators, specifically addressing the `@foo` vs `@foo()` issue.

willronchetti reviewed (approved):
LGTM, just needs version bump

netsettler commented on dcicutils/misc_utils.py:
```diff
@@ -801,6 +820,12 @@ def check_true(test_value, message, error_class=None):
         raise error_class(message)
 
 
+def reversed(seq):
+    seq_copied = list(seq).copy()
+    seq_copied.reverse()
+    return seq_copied
```
Weird. I looked for a builtin and I swear it was not there! I can live with the builtin given it's there.

netsettler commented on dcicutils/misc_utils.py:
```diff
@@ -889,11 +929,53 @@ def full_object_name(obj):
-    def fn(*args, **kwargs):
+class _ConstantValuedFunction:
+
+    def __init__(self, value):
+        self._value = value
+
+    def __str__(self):
+        return repr(self)
+
+    def __repr__(self):
+        return "constantly(%r)" % self._value
+
+    def __call__(self, *args, **kwargs):
         ignored(args, kwargs)
-        return value
-    return fn
+        return self._value
+
+    @property
+    def __doc__(self):
+        return "A function that always returns a constant value: %r" % self._value
```
(Yeah, it's just to help in debugging so if you see the function you can know what it is.)

netsettler commented:
In response to comments by @willronchetti, I did this:

* Removed `reversed`. I can't believe I didn't see that was a built-in. I would swear I checked, even under both spellings. Anyway, I'm happy to use the built-in.
* Removed the reimplementation of `constantly`. Will had asked what this was fixing, and I was going to say efficiency, because I'd just assumed it was simpler. Python's notion of compilation is weird, though, and there's some overhead in instances, I guess, because it's just not as fast. So I just backed it out. What's there is fine.

I adjusted the initial PR comment so it would match what I'm submitting. I bumped the version and added a changelog entry.

netsettler commented:
Unit tests seem to be passing. Something must have changed on mastertest. The tests are presumably still fragile, but I'll take the win for now and merge this quick before it changes its mind. The failing tests had nothing to do with these changes, I just hate merging with a failing test.
